File: apps/web/lib/api/notifications.ts
import axios from 'axios'
import { API_BASE_URL } from '../config'

export interface Notification {
  id: string
  user_id?: string
  tenant_id: string
  type: string
  title: string
  message: string
  entity_type?: string
  entity_id?: string
  read: boolean
  created_at: string
}

export interface NotificationCount {
  count: number
}

export const notificationApi = {
  list: async (read?: boolean, limit: number = 50, offset: number = 0): Promise<Notification[]> => {
    const token = localStorage.getItem('token')
    const headers = token ? { 'Authorization': `Bearer ${token}` } : {}
    const params = new URLSearchParams()
    if (read !== undefined) params.append('read', String(read))
    params.append('limit', String(limit))
    params.append('offset', String(offset))
    
    const res = await axios.get(`${API_BASE_URL}/notifications?${params}`, {
      headers,
      withCredentials: true
    })
    return res.data
  },

  getUnreadCount: async (): Promise<number> => {
    const token = localStorage.getItem('token')
    const headers = token ? { 'Authorization': `Bearer ${token}` } : {}
    const res = await axios.get(`${API_BASE_URL}/notifications/unread-count`, {
      headers,
      withCredentials: true
    })
    return res.data.count
  },

  markRead: async (notificationId: string): Promise<Notification> => {
    const token = localStorage.getItem('token')
    const headers = token ? { 'Authorization': `Bearer ${token}` } : {}
    const res = await axios.post(`${API_BASE_URL}/notifications/${notificationId}/read`, {}, {
      headers,
      withCredentials: true
    })
    return res.data
  },

  markAllRead: async (): Promise<{ marked_read: number }> => {
    const token = localStorage.getItem('token')
    const headers = token ? { 'Authorization': `Bearer ${token}` } : {}
    const res = await axios.post(`${API_BASE_URL}/notifications/read-all`, {}, {
      headers,
      withCredentials: true
    })
    return res.data
  }
}

File: apps/api/app/api/notifications.py
"""
Notification endpoints for in-app alerts
"""
from fastapi import APIRouter, Depends, Query, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, or_
from typing import Optional, List
from uuid import UUID
from datetime import datetime, timezone

from ..db.session import get_session
from .. import models, schemas
from ..core.auth import get_current_user, CurrentUser

router = APIRouter()


@router.get("", response_model=List[schemas.NotificationOut])
async def list_notifications(
    read: Optional[bool] = Query(None, description="Filter by read status"),
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_session),
    current_user: CurrentUser = Depends(get_current_user)
):
    """List notifications for current user"""
    query = select(models.Notification).where(
        or_(
            models.Notification.user_id == current_user.id,
            and_(
                models.Notification.user_id.is_(None),
                models.Notification.tenant_id == current_user.tenant_id
            )
        )
    )
    
    if read is not None:
        query = query.where(models.Notification.read == read)
    
    query = query.order_by(models.Notification.created_at.desc())
    query = query.limit(limit).offset(offset)
    
    result = await db.execute(query)
    notifications = result.scalars().all()
    
    return notifications


@router.get("/unread-count", response_model=schemas.NotificationCount)
async def get_unread_count(
    db: AsyncSession = Depends(get_session),
    current_user: CurrentUser = Depends(get_current_user)
):
    """Get count of unread notifications for current user"""
    query = select(models.Notification).where(
        and_(
            or_(
                models.Notification.user_id == current_user.id,
                and_(
                    models.Notification.user_id.is_(None),
                    models.Notification.tenant_id == current_user.tenant_id
                )
            ),
            models.Notification.read == False
        )
    )
    
    result = await db.execute(query)
    count = len(result.scalars().all())
    
    return {"count": count}


@router.post("/{notification_id}/read", response_model=schemas.NotificationOut)
async def mark_notification_read(
    notification_id: UUID,
    db: AsyncSession = Depends(get_session),
    current_user: CurrentUser = Depends(get_current_user)
):
    """Mark a notification as read"""
    result = await db.execute(
        select(models.Notification).where(
            and_(
                models.Notification.id == notification_id,
                or_(
                    models.Notification.user_id == current_user.id,
                    and_(
                        models.Notification.user_id.is_(None),
                        models.Notification.tenant_id == current_user.tenant_id
                    )
                )
            )
        )
    )
    notification = result.scalar_one_or_none()
    
    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")
    
    notification.read = True
    await db.commit()
    await db.refresh(notification)
    
    return notification


@router.post("/read-all")
async def mark_all_read(
    db: AsyncSession = Depends(get_session),
    current_user: CurrentUser = Depends(get_current_user)
):
    """Mark all notifications as read for current user"""
    query = select(models.Notification).where(
        and_(
            or_(
                models.Notification.user_id == current_user.id,
                and_(
                    models.Notification.user_id.is_(None),
                    models.Notification.tenant_id == current_user.tenant_id
                )
            ),
            models.Notification.read == False
        )
    )
    
    result = await db.execute(query)
    notifications = result.scalars().all()
    
    for notification in notifications:
        notification.read = True
    
    await db.commit()
    
    return {"marked_read": len(notifications)}

